src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -196,6 +198,138 @@ public void GenerateHeader()
+
+            var method = function as Method;
+            var isInstanceMethod = method != null && !method.IsStatic;
+
+            if (isInstanceMethod && Options.IsMicrosoftAbi)
+            {
+                @params.Add("global::System.IntPtr instance");
+
+                if (method.IsConstructor)
+                    retType = "global::System.IntPtr";
+            }
+
+            if (!function.HasIndirectReturnTypeParameter &&
+                isInstanceMethod && Options.IsItaniumLikeAbi)
+                @params.Add("global::System.IntPtr instance");
+
+            int i = 0;
+            foreach (var param in function.Parameters)
+            {
+                if (param.Kind == ParameterKind.OperatorParameter)
+                    continue;
+
+                var typeName = param.CSharpType(TypePrinter);
+
+                @params.Add(string.Format("{0} param{1}", typeName, i++));
+
+                if (param.Kind == ParameterKind.IndirectReturnType &&
+                    isInstanceMethod && Options.IsItaniumLikeAbi)
+                    @params.Add("global::System.IntPtr instance");
+            }
+
+            if (method != null && method.IsConstructor)
+            {
+                var @class = (Class)method.Namespace;
+                if (Options.IsMicrosoftAbi && @class.Layout.HasVirtualBases)
+                    @params.Add("int " + GeneratedIdentifier("forBases"));
+            }
+
+            TypePrinter.PopContext();
+
+            return string.Join(", ", @params);
+        }
+
+        private Dictionary<DelegateSignature, string> AllDelegates = new Dictionary<DelegateSignature, string>();
+
+        private string GenerateDelegate(Method func, bool suppressWarning = true)
+        {
+            DelegateSignature ds = new DelegateSignature();
+            ds.callingConvention = func.CallingConvention.ToInteropCallConv();
+            CSharpTypePrinterResult retType;
+            ds.signature = GenerateDelegateSignature(func, out retType);
+            ds.retType = retType.ToString();
+            ds.suppressWarning = suppressWarning;
+
+            string delegateName;
+            if (AllDelegates.TryGetValue(ds, out delegateName))
+                return "DelegatesStorage." + delegateName;
+
+            if (ds.retType.Equals("void"))
+                delegateName = "Func_" + AllDelegates.Count;
+            else
+                delegateName = "Action_" + AllDelegates.Count;
+            AllDelegates.Add(ds, delegateName);
+
+            return "DelegatesStorage." + delegateName;
+        }
+        private string GenerateDelegate(Function func, bool suppressWarning = true)
+        {
+            return GenerateDelegate(func as Method, suppressWarning);
+        }
+
+        private void GenerateDelegatesClass()
+        {
+            if (AllDelegates.Count == 0)
+                return;
+            WriteLine("");
+            WriteLine("public unsafe class DelegatesStorage");
```

Comment:
I'd rather not have the delegates in a class because people would wonder what the class itself is for. Please generate a name-space instead, or place them directly in the root (without a name-space) - I am not sure at the moment which would be better.